### DISCUSSION TITLE
c# M2mqtt to AWS IoT c++

### DISCUSSION THREAD

**Discussion by autoprog** [aws/aws-iot-device-sdk-cpp-v2#513] [General]
I have a C# M2Mqtt application that works perfect on IoT AWS.

c# 
```c#
 var clientCert = new X509Certificate2(Path.Combine(AppContext.BaseDirectory, "certificate.pfx"), @"");
 var caCert = X509Certificate.CreateFromCertFile(Path.Combine(AppContext.BaseDirectory, "root.pem"));
 client = new MqttClient(IotEndpoint, BrokerPort, true, caCert, clientCert, MqttSslProtocols.TLSv1_2);
 client.MqttMsgPublishReceived += client_MqttMsgPublishReceived;
 client.ProtocolVersion = MqttProtocolVersion.Version_3_1;
 client.Connect(device_aws_id);
 client.Subscribe(new[] { "topic" }, new[] { MqttMsgBase.QOS_LEVEL_AT_LEAST_ONCE });
```

my file certificate.pfx ok
my file root.pem ok
//-------------------------

but I need to implement Mqtt in a c++ library.

sdk v2 setup ok,
C:\dev\iotsdk\sdk-cpp-workspace ok
use Visual Studio 2019, it see sdk ok

Does any partner have an example that works.
which example looks like my code.

thanks,
Carlos.

**Comment by jmklix** [comment]
Have you taken a look at our [x509_credentials_provider_connect](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/main/samples/mqtt/x509_credentials_provider_connect/main.cpp) sample? It shows how to connect in c++ with x509 credentials

**Reply by autoprog** [reply]
hi,
jmklix, thanks,
my test, error link2019

Error LNK2019 unresolved external symbol, _aws_mem_acquire, referenced in function "public: char * __thiscall Aws::Crt::StlAllocator<char>::allocate(unsigned int,void const *)" 
(?allocate@?$StlAllocator@D@ Crt@Aws@@QAEPADIPBX@Z) 
AWSMqttc D:\..AWSMqttc.obj 1

![image](https://user-images.githubusercontent.com/5191545/211906702-43b918c2-50d2-4339-a487-840f3d5dd9cc.png)

trying to solve,

thanks,
Carlos.

**Comment by bretambrose** [comment]
The code sample in the original query is just using certificate and private key to connect.  The X509 credentials provider is not what they'd want here.
